src/modules/cpu/kernel/transpose_generic.hpp

```diff
@@ -0,0 +1,441 @@
+{
+    Rpp32u dstShape[RPPT_MAX_DIMS];
+    Rpp32u srcStrides[RPPT_MAX_DIMS];
+    Rpp32u dstStrides[RPPT_MAX_DIMS];
+
+    // compute output shape
+    for(Rpp32u i = 0; i < nDim; i++)
+        dstShape[i] = length[perm[i]];
+
+    // compute output strides
+    compute_strides(dstStrides, dstShape, nDim);
+
+    // compute input strides and update as per the permute order
+    Rpp32u tempStrides[RPPT_MAX_DIMS];
+    compute_strides(tempStrides, length, nDim);
+    for(int i = 0; i < nDim; i++)
+        srcStrides[i] = tempStrides[perm[i]];
+
+    // perform transpose as per the permute order
+    transpose_generic_nd_recursive(dstPtrTemp, dstStrides, srcPtrTemp, srcStrides, dstShape, nDim);
+}
+
+RppStatus transpose_generic_f32_f32_host_tensor(Rpp32f *srcPtr,
+                                                RpptGenericDescPtr srcGenericDescPtr,
+                                                Rpp32f *dstPtr,
+                                                RpptGenericDescPtr dstGenericDescPtr,
+                                                Rpp32u *permTensor,
+                                                Rpp32u *roiTensor,
+                                                RppLayoutParams layoutParams,
+                                                rpp::Handle& handle)
+{
+    Rpp32u numThreads = handle.GetNumThreads();
+    Rpp32u nDim = dstGenericDescPtr->numDims;
+    Rpp32u batchSize = dstGenericDescPtr->dims[0];
+
+    omp_set_dynamic(0);
+#pragma omp parallel for num_threads(numThreads)
+    for(int batchCount = 0; batchCount < batchSize; batchCount++)
+    {
+        Rpp32f *srcPtrTemp, *dstPtrTemp;
+        srcPtrTemp = srcPtr + batchCount * srcGenericDescPtr->strides[0];
+        dstPtrTemp = dstPtr + batchCount * dstGenericDescPtr->strides[0];
+
+        // get the starting address of begin and length values from roiTensor
+        Rpp32u *roi = roiTensor + batchCount * nDim * 2;
+        Rpp32u *begin = roi;
+        Rpp32u *length = &roi[nDim];
+        Rpp32u *perm = permTensor;
+
+        bool copyInput = true;
+        for(int i = 0; i < nDim; i++)
+            copyInput *= (perm[i] == i);
+
+        // do memcpy of input to output since output order is same as input order
+        if(copyInput)
+        {
+            memcpy(dstPtrTemp, srcPtrTemp, (size_t)(srcGenericDescPtr->strides[0] * sizeof(Rpp32f)));
+        }
+        else
+        {
+            for(int i = 1; i < nDim; i++)
+                srcPtrTemp += begin[i] * srcGenericDescPtr->strides[i];
+
+            if (nDim == 2 && perm[0] == 1 && perm[1] == 0)
+            {
+                compute_2d_transpose(srcPtrTemp, dstPtrTemp, length[0], length[1], srcGenericDescPtr->strides[1], dstGenericDescPtr->strides[1]);
+            }
+            else if (nDim == 3)
+            {
+                if(perm[0] == 2 && perm[1] == 0 && perm[2] == 1 && length[2] == 16)
```

Comment:
Wherever you have a more optimized version, state it as a comment above the if statement. // Optimized AVX version for <>dims or for <>permutation